### ISSUE TITLE
Focus trap breaks clipboard.js implementations within a trap

### ISSUE THREAD

**Issue by DesignerDave** [focus-trap/focus-trap#383]
This issue seems like a duplicate of https://github.com/focus-trap/focus-trap/issues/91, though that issue was closed before a more holistic solution was reached.

I work on a component library, and we use focus trap to help make our modal components more accessible. Unfortunately, the implementation of focus trap seems incompatible with clipboard.js. As noted in the issue above, there is a workaround that can be applied on a case-by-case basis, but since we are building a component library, this is creating a downstream impact on developers working with our modals with clipboard.js.

Here's an example using our own Copyable component with a modal that implements focus trap: https://codesandbox.io/s/inspiring-nightingale-vi3jp?file=/src/App.tsx. The text within the modal isn't copied when the copy button is clicked, but the text outside the modal works as expected.

**Comment by stefcameron** [comment]
@DesignerDave Thanks for the repro!

So did you try this `data-clipboard-target` (or is it `data-clipboard-text`, not sure, author says both while seeming to refer to them as the same thing) https://github.com/focus-trap/focus-trap/issues/91#issuecomment-544991383 ?

If not that, what if you used `focus-trap`'s relatively new ability to work with multiple containers? At the moment when you need to copy the text and clipboard.js is about to create its element it needs to set focus to temporarily, if that element was created in a container (could be a container separate from the modal), then clipboard.js should then be able to set focus to it in order to trigger the browser to copy the selected text. You can update containers on-the-fly even after the trap has been created with the new [updateContainerElements9)](https://github.com/focus-trap/focus-trap#trapupdatecontainerelements) API.

**Comment by DesignerDave** [comment]
Unless I'm missing something, the `data-clipboard-` attribute approach is a solution to solving the issue where we control the clipboard.js implementation (e.g. our Copyable component used in the codesandbox case I sent), but doesn't solve for the case where someone using our Modal component wants to use the clipboard library directly. Please let me know if I'm missing something in that respect.

The `updateContainerElements` API approach might let us do something interesting here. I appreciate the response – this might make it possible to do what we need to.

The challenge I see here is understanding when we need to update the focus trap container when we aren't aware of when the clipboard.js trigger itself is clicked. Perhaps we could create an event on descendants of the focus-trap with the `data-clipboard-text` attribute, and change the focus-trap container then? Seems more brittle than ideal, but it might be the only option unless you have another idea 🤔

**Comment by stefcameron** [comment]
> Unless I'm missing something, the `data-clipboard-` attribute approach is a solution to solving the issue where we control the clipboard.js implementation (e.g. our Copyable component used in the codesandbox case I sent), but doesn't solve for the case where someone using our Modal component wants to use the clipboard library directly. Please let me know if I'm missing something in that respect.

That's possible. But whether it's your Modal that's doing it automatically behind the scenes, or it's a consumer of your Modal doing it by themselves, it's possible this can help fix the problem.

> The `updateContainerElements` API approach might let us do something interesting here. I appreciate the response – this might make it possible to do what we need to.
> 
> The challenge I see here is understanding when we need to update the focus trap container when we aren't aware of when the clipboard.js trigger itself is clicked. Perhaps we could create an event on descendants of the focus-trap with the `data-clipboard-text` attribute, and change the focus-trap container then? Seems more brittle than ideal, but it might be the only option unless you have another idea 🤔
> 
> Really appreciate the quick response by the way!

My pleasure! Looks like you'd have to cover `data-clipboard-target`, `data-clipboard-action` and `data-clipboard-text` as triggers for clipboard.js to do its thing. Seems like, as you say, you could try `querySelectorAll()` with a selector to get all elements with those attributes and add click event handlers to them. There might be an issue with timing, though, since I'll bet clipboard.js's handlers get added before yours, and then by the time your handler gets called, clipboard.js may have already failed.

Not sure if this would meet your library's minimum supported browser matrix, but what if you used a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to look for the temporary element that clipboard.js creates? Just a thought I didn't think too long about but maybe worth digging into. Still would be brittle because what clipboard.js creates is an implementation detail.

Maybe they could provide an enhancement that gives you a better ability to hook into what they're doing, since putting clipboard.js into a focus-trap seems like a perfectly fine use case. If the library couldn't just be directed to create that temporary element within a predefined container that you could then give to focus-trap when you first create the trap, I think you'd be set. This container could be 1x1 pixel off screen somewhere, etc.

**Comment by ryantheleach** [comment]
Seems to me that it's failing not because of the focus trap, but because the method used by Clipboard.js seems to be a hacky remnant / case of compatibility with ancient browsers. Adding the temporary input next to the text being copied / button enacting the copy would work as it'd now be inside the focus-trap, instead of adding it to the document, as would using a Range object to select the text in the paragraph being displayed to the user.

Basically, focus-trap is doing it's job, preventing focus to an input outside the focus trap.

**Comment by ryantheleach** [comment]
After searching into the issue for a few hours, it's also probably also down to the way focus-trap emulates the browsers native tab order.

https://github.com/vue-a11y/vue-focus-loop/ relies on a different method of trapping focus, in that it waits for the focus to enter one of the sentinel tabbable panels before and after the target locked element.

Because of this, I suspect clipboard.js *might* work there, however unless clipboard.js returns focus, there could be a whole new issue of focus escaping the lock.

**Comment by stefcameron** [comment]
> https://github.com/vue-a11y/vue-focus-loop/ relies on a different method of trapping focus, in that it waits for the focus to enter one of the sentinel tabbable panels before and after the target locked element.

Thanks for doing some research. `focus-trap` relies on `tabbable` to determine what nodes are tabbable/focusable inside any container given to the trap (multiple containers are supported, `focus-trap` will cycle through them in order given). `tabbable` does infer an order according to web standards. `focus-trap` then mostly uses that order to know what each container's boundary nodes are (first and last tabbable nodes). Otherwise, it gets out of the way and lets the browser dictate order for nodes that are natively tabbable and ones that have `tabindex="0"`. Nodes that have `tabindex` > 0 are a special case, and I think focus-trap will need to enforce that order as best it can, see https://github.com/focus-trap/focus-trap/issues/375

**Comment by stefcameron** [comment]
I just found https://www.npmjs.com/package/copy-text-to-clipboard, which has this `target` option: "Specify a DOM element where the temporary, behind-the-scenes textarea should be appended, in cases where you need to stay __within a focus trap__, like in a modal." (emphasis mine).

I think you're tied to Clipboard.js, but it seems this is a common problem, and solvable, albeit from the inside...

**Comment by stefcameron** [comment]
Closing very old issues. See my most recent comment from a ~year ago. If it's still an issue with the current release, feel free to re-open.
